Provide drush command and auth fallback to generate tokens

snwalkunde opened the pull request:
Provide drush command and auth fallback to generate tokens
- https://www.drupal.org/project/eloqua_api_redux/issues/3053639
- Implement resource owner password grant authentication to generate access and refresh tokens and provide drush command for the same using auth fallback module
- Implement fallback for gerenrate access and refresh tokens when they are expired/invalidate

snwalkunde committed:
Fix - https://www.drupal.org/project/eloqua_api_redux/issues/3053639

- Implement resource owner password grant authentication to generate access and refresh tokens and provide drush command for the same using auth fallback module
- Implement fallback for gerenrate access and refresh tokens when they are expired/invalidate

snwalkunde committed:
Fix - https://www.drupal.org/project/eloqua_api_redux/issues/3053639

- Implement resource owner password grant authentication to generate access and refresh tokens and provide drush command for the same using auth fallback module
- Implement fallback for gerenrate access and refresh tokens when they are expired/invalidate

snwalkunde committed:
Merge branch 'eloqua_auth_fallback' of https://github.com/dakkusingh/eloqua_api_redux into eloqua_auth_fallback

snwalkunde commented on src/Service/EloquaApiClient.php:
```diff
@@ -150,7 +162,7 @@ public function getAccessTokenByRefreshToken() {
    *   new refresh token
    */
-  private function doTokenRequest(array $params, $res = 'token') {
+  public function doTokenRequest(array $params, $res = 'token') {
```
This was changed deliberately so that `doTokenRequest` can be reused in drush service

dakkusingh replied:
private to public makes sense :)

dakkusingh commented on modules/eloqua_api_auth_fallback/config/schema/eloqua_api_auth_fallback.schema.yml:
```diff
@@ -0,0 +1,13 @@
+eloqua_api_auth_fallback.settings:
+  type: config_object
+  label: 'Eloqua API Authentication settings'
```
can we name this to something a little more descriptive? maybe:
"**Eloqua API resource owner password credentials**" or something reflecting this?

snwalkunde replied:
Yes. I have updated the code to incorporate this change.

dakkusingh commented on modules/eloqua_api_auth_fallback/eloqua_api_auth_fallback.links.menu.yml:
```diff
@@ -0,0 +1,5 @@
+eloqua_api_auth_fallback.settings:
+  title: 'Eloqua API Authentication Settings'
```
can we name this to something a little more descriptive? maybe:
"Eloqua API resource owner password credentials" or something reflecting this?

snwalkunde replied:
Done

dakkusingh commented on modules/eloqua_api_auth_fallback/eloqua_api_auth_fallback.links.menu.yml:
```diff
@@ -0,0 +1,5 @@
+eloqua_api_auth_fallback.settings:
+  title: 'Eloqua API Authentication Settings'
+  route_name: eloqua_api_redux.settings
+  description: 'Configure Eloqua API Authentication Settings.'
```
can we name this to something a little more descriptive? maybe:
"Configure Eloqua API resource owner password credentials" or something reflecting this?

snwalkunde replied:
Done

dakkusingh commented on modules/eloqua_api_auth_fallback/eloqua_api_auth_fallback.routing.yml:
```diff
@@ -0,0 +1,7 @@
+  path: 'admin/config/services/eloqua_api_redux/auth_settings'
+  defaults:
+    _form:  '\Drupal\eloqua_api_auth_fallback\Form\Settings'
+    _title: 'Eloqua API Authentication Settings'
```
same as above feedback for desc.

snwalkunde replied:
Done

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Form/Settings.php:
```diff
@@ -0,0 +1,72 @@
+
+    $form['credentials'] = [
+      '#type' => 'fieldset',
+      '#title' => $this->t('Eloqua Instance Authentication Credentials'),
```
same as above feedback for desc.

dakkusingh replied:
Given all 3 values are required as per the [API docs](https://docs.oracle.com/cloud/latest/marketingcs_gs/OMCAB/Developers/GettingStarted/Authentication/authenticate-using-oauth.htm#resource-owner-password-credentials-grant)..
might be a good idea to set `'#required' => TRUE`

snwalkunde replied:
Done

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Form/Settings.php:
```diff
@@ -0,0 +1,72 @@
+    $form['credentials']['username'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Username'),
+      '#default_value' => $config->get('username'),
```
Might be a good idea to add a `#description`

dakkusingh replied:
Given all 3 values are required as per the [API docs](https://docs.oracle.com/cloud/latest/marketingcs_gs/OMCAB/Developers/GettingStarted/Authentication/authenticate-using-oauth.htm#resource-owner-password-credentials-grant)..
might be a good idea to set `'#required' => TRUE`

snwalkunde replied:
I have set required for these fields. Missed the description part will add it shortly

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Form/Settings.php:
```diff
@@ -0,0 +1,72 @@
+    $form['credentials']['sitename'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Instance/Site Name'),
+      '#default_value' => $config->get('sitename'),
```
Might be a good idea to add a `#description`

dakkusingh replied:
Given all 3 values are required as per the [API docs](https://docs.oracle.com/cloud/latest/marketingcs_gs/OMCAB/Developers/GettingStarted/Authentication/authenticate-using-oauth.htm#resource-owner-password-credentials-grant)..
might be a good idea to set `'#required' => TRUE`

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Form/Settings.php:
```diff
@@ -0,0 +1,72 @@
+    $form['credentials']['password'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Password'),
+      '#default_value' => $config->get('password'),
```
Might be a good idea to add a `#description`

dakkusingh replied:
Given all 3 values are required as per the [API docs](https://docs.oracle.com/cloud/latest/marketingcs_gs/OMCAB/Developers/GettingStarted/Authentication/authenticate-using-oauth.htm#resource-owner-password-credentials-grant)..
might be a good idea to set `'#required' => TRUE`

dakkusingh commented on src/Service/EloquaApiClient.php:
```diff
@@ -131,6 +131,18 @@ public function getAccessTokenByRefreshToken() {
+      // password grant authorization.
+      // Also make sure that auth fallback service exists i.e. module that
+      // holds this service is enabled.
+      if (\Drupal::hasService('eloqua_api_auth_fallback.commands')) {
```
It would be better if this is Dependency Injected. See
[Dependency Injection if service exists](https://drupal.stackexchange.com/questions/273178/dependency-injection-if-service-exists)

snwalkunde replied:
I tried to inject the auth fallback service using service provider approach however since there is dependency already added in auth fallback service for eloqua client, system ran into circular reference error with following message

> Symfony\Component\DependencyInjection\Exception\ServiceCircularReferenceException: Circular reference detected for service "eloqua_api_redux.client", path: "eloqua_api_redux.forms -> eloqua_api_redux.client -> eloqua_api_redux.auth_fallback_default". in Drupal\Component\DependencyInjection\Container->get() (line 141 of /var/www/rmld8/web/core/lib/Drupal/Component/DependencyInjection/Container.php).

Tried following approaches but still couldn't resolve this issue.
1. Lowering the dependency priorities
2. Declaring service to be loaded with lazy 

Below are some of the similar drupal org issues 
- https://www.drupal.org/project/domain_language/issues/3007609
- https://www.drupal.org/project/domain/issues/3005048
- https://www.drupal.org/project/amp/issues/3050965

Had to resort with [static loading of the auth fallback service](https://github.com/dakkusingh/eloqua_api_redux/pull/1/commits/db9c112b7b03869418e694dac1c35d57d36e982d#diff-bf1636aaa4e03b2e3f702b491e26a061R138). 


I have implemented default auth fallback service and overridden it with submodule service using service decoration pattern. I felt this approach is clean. Let me know if this fine else will revert to earlier implementation.

dakkusingh replied:
> overridden it with submodule

At a glance that looks quite neat as well.

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Commands/EloquaAuthTokensGenerate.php:
```diff
@@ -0,0 +1,104 @@
+    $params = [
+      'grant_type' => 'password',
+      'scope' => 'full',
+      'username' => $sitename . '\\' . $username,
```
is `\\` the correct separator? 
As per the [API docs](https://docs.oracle.com/cloud/latest/marketingcs_gs/OMCAB/Developers/GettingStarted/Authentication/authenticate-using-oauth.htm#resource-owner-password-credentials-grant) isnt it `sitename` + `/` + `username`?

snwalkunde replied:
Yes \\ is correct separator. This is present in [API docs example](https://docs.oracle.com/cloud/latest/marketingcs_gs/OMCAB/Developers/GettingStarted/Authentication/authenticate-using-oauth.htm#resource-owner-password-credentials-grant)

`sitename + / + username?` isnt working correctly and resulting in 400 response 

>  [error]  Client error: POST  resulted in a `400 Bad Request` response:
{"error":"invalid_grant","error_description":"The site, username, or password are invalid."}

dakkusingh replied:
Wrong API docs from Oracle? Not surprised :)

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Commands/EloquaAuthTokensGenerate.php:
```diff
@@ -0,0 +1,104 @@
+use Drush\Commands\DrushCommands;
+
+/**
+ * A drush command for eloqua api authentication module.
```
See above comment for writing better description!

snwalkunde replied:
Updated

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Commands/EloquaAuthTokensGenerate.php:
```diff
@@ -0,0 +1,104 @@
+   *   Calls eloqua authentication API and generate tokens
+   *   using Resource Owner Password Credentials grant.
+   */
+  public function generateTokens() {
```
might be a good idea to make this method name more descriptive (`generateTokensByResourceOwner`) so alternative methods can be added to the same class later on.

snwalkunde replied:
Done

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Commands/EloquaAuthTokensGenerate.php:
```diff
@@ -0,0 +1,104 @@
+    $sitename = $this->configSettings->get('sitename');
+    $username = $this->configSettings->get('username');
+    $password = $this->configSettings->get('password');
+    if (empty($sitename) || empty($username) || empty($password)) {
```
this is not required if you make the form fields mandatory in the settings.php

snwalkunde replied:
I think its safe/defensive to keep the empty checks here. We would avoid API call failures if module is installed for someone forgets to set configuration credentials

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Commands/EloquaAuthTokensGenerate.php:
```diff
@@ -0,0 +1,104 @@
+      $message = 'Error updating refresh and access tokens.';
+      $this->output()
+        ->writeln($message);
+      $this->loggerFactory->get('eloqua_api_auth_fallback')->error($message);
```
[EloquaApiClient::doTokenRequest() already has logging on failed request](https://github.com/dakkusingh/eloqua_api_redux/blob/7e0752e7867ccf2062f51de989a5ee803b27bceb/src/Service/EloquaApiClient.php#L202). I dont think log duplication is required here.

snwalkunde replied:
Removed

snwalkunde committed:
Fix - https://www.drupal.org/project/eloqua_api_redux/issues/3053639

- Improve comments, field titles and descriptions
- Implement default token provider and service decorator
- Make credential fields required on configuration settings
- Fix auth configuration link which was ealier pointing to parent route

dakkusingh commented on src/Service/EloquaAuthDefaultFallback.php:
```diff
@@ -0,0 +1,19 @@
+namespace Drupal\eloqua_api_redux\Service;
+
+/**
+ * Interface for migrations.
```
copy paste typo?

snwalkunde replied:
Fixed this

dakkusingh commented on src/Service/EloquaAuthFallbackInterface.php:
```diff
@@ -0,0 +1,21 @@
+namespace Drupal\eloqua_api_redux\Service;
+
+/**
+ * Interface for migrations.
```
copy paste typo "migrations"?

snwalkunde replied:
Fixed.

snwalkunde committed:
- Add descriptions to auth credentials settings fields
- Fix comments in file

snwalkunde committed:
Fix interface and class file comments

dakkusingh commented on modules/eloqua_api_auth_fallback/src/Form/Settings.php:
```diff
@@ -0,0 +1,89 @@
+      '#type' => '#markup',
+      '#markup' => $this->t('For more information on authenticate using a resource owner password credentials grant, refer this @link.',
+        [
+          '@link' => Link::fromTextAndUrl('https://docs.oracle.com/cloud/latest/marketingcs_gs/OMCAB/Developers/GettingStarted/Authentication/authenticate-using-oauth.htm#resource-owner-password-credentials-grant',
```
Can this be text like "**refer to API Documentation for resource owner password credentials grant**" with hyperlink?

snwalkunde replied:
Link updated

snwalkunde committed:
- Improve help links for eloqua documentation

dakkusingh reviewed (approved):
looks good. please go ahead and merge after PR2.

snwalkunde committed:
Merge branch '8.x-1.x' into eloqua_auth_fallback
and Resolve conflicts
